plutus-core/plutus-core/src/PlutusCore/Builtin/Polymorphism.hs

```diff
@@ -0,0 +1,252 @@
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE KindSignatures        #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE PolyKinds             #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+{-# LANGUAGE UndecidableInstances  #-}
+
+module PlutusCore.Builtin.Polymorphism
+    ( Opaque (..)
+    , SomeConstant (..)
+    , TyNameRep (..)
+    , TyVarRep
+    , TyAppRep
+    , TyForallRep
+    ) where
+
+import PlutusCore.Builtin.HasConstant
+import PlutusCore.Core
+import PlutusCore.Pretty
+
+import Data.Kind qualified as GHC (Type)
+import GHC.Ix
+import GHC.TypeLits
+import Universe
+
+{- Note [Motivation for polymorphic built-in functions]
+We need to support polymorphism for built-in functions for these reasons:
+
+1. @ifThenElse@ for 'Bool' (being a built-in type rather than a Scott-encoded one) has to be
+   polymorphic as its type signature is
+
+       ifThenElse : all a. Bool -> a -> a -> a
+
+   Previously we had 'Bool' as a Scott-encoded type, but this required plenty of supporting machinery,
+   because unlifting (aka Scott-decoding) a PLC 'Bool' into a Haskell 'Bool' is quite a non-trivial
+   thing, see https://github.com/input-output-hk/plutus/blob/e222466e6d46bbca9f76243bb496b3c88ed02ca1/language-plutus-core/src/PlutusCore.Builtin/Typed.hs#L165-L252
+
+   Now that we got rid of all this complexity we have to pay for that by supporting polymorphic
+   built-in functions (but we added that support long ago anyway, 'cause it was easy to do).
+
+2. we may want to add efficient polymorphic built-in types like @IntMap@ or @Vector@ and most functions
+   defined over them are polymorphic as well
+-}
+
+-- See Note [Motivation for polymorphic built-in functions].
+-- See Note [Implementation of polymorphic built-in functions].
+-- See Note [Pattern matching on built-in types].
+-- | The denotation of a term whose PLC type is encoded in @rep@ (for example a type variable or
+-- an application of a type variable). I.e. the denotation of such a term is the term itself.
+-- This is because we have parametricity in Haskell, so we can't inspect a value whose
+-- type is, say, a bound variable, so we never need to convert such a term from Plutus Core to
+-- Haskell and back and instead can keep it intact.
+newtype Opaque val (rep :: GHC.Type) = Opaque
+    { unOpaque :: val
+    } deriving newtype (Pretty, AsConstant, FromConstant)
+
+type instance UniOf (Opaque val rep) = UniOf val
+
+-- | For unlifting from the 'Constant' constructor when the stored value is of a monomorphic
+-- built-in type
+--
+-- The @rep@ parameter specifies how the type looks on the PLC side (i.e. just like with
+-- @Opaque val rep@).
+newtype SomeConstant uni (rep :: GHC.Type) = SomeConstant
```

Comment:
This isn't to do with polymorphism, this is just how we represent constants. Maybe should be somewhere else?

Reply:
No, it's literally for polymorphism. Note how similar it is to `Opaque`. It only makes sense to use `SomeConstant` when you have an argument whose type has type variables in it (or an even fancier Plutus type). Check out how `SomeConstant` is used in the `Default.Builtins` and the `Examples.Builtins` modules.

Reply:
Ugh, I'm getting confused with `ValueOf`.